coding-style.md

```diff
@@ -0,0 +1,552 @@
+
+- The `const` keyword precedes the type, e.g., `const T&` as opposed to
+  `T const&`.
+
+- `*` and `&` bind to the *type*, e.g., `T* arg` instead of `T *arg`.
+
+- When declaring variables and functions, provide the [storage class
+  specifier] (`extern`, `static`, `thread_local`, `mutable`) first, followed
+  by the [declaration specifiers] in order of `friend`, `inline`, `virtual`,
+  `explicit`, `constexpr`, `consteval`, and `constinit`.
+
+[storage class specifier]: https://en.cppreference.com/w/cpp/language/storage_duration
+[declaration specifiers]: https://en.cppreference.com/w/cpp/language/declarations#Specifiers
+
+- Always use `auto` to declare a variable unless you cannot initialize it
+  immediately or if you actually want a type conversion. In the latter case,
+  provide a comment why this conversion is necessary.
+
+- Never use unwrapped, manual resource management such as `new` and `delete`.
+
+- Never use `typedef`; always write `using T = X` in favor of `typedef X T`.
+
+- Keywords are always followed by a whitespace: `if (...)`, `template <...>`,
+  `while (...)`, etc.
+
+- Do not add whitespace when negating an expression with `!`:
+
+  ```cpp
+  if (!sunny())
+    stay_home()
+  ```
+
+- Opening braces belong onto the same line:
+
+  ```cpp
+  struct foo {
+    void f() {
+    }
+  };
+  ```
+
+- Use inline functions for trivial code, such as getters/setters or
+  straight-forward logic that does not span more than 3 lines.
+
+### Header
+
+- Header filenames end in `.hpp` and implementation filenames in `.cpp`.
+
+- All header files should use `#pragma once` to prevent multiple inclusion.
+
+- Don't use `#include` when a forward declarations suffices. It can make sense
+  to outsource forward declarations into a separate file per module. The file
+  name should be `<MODULE>/fwd.h`.
+
+- Include order is from high-level to low-level headers, e.g.,
+
+  ```cpp
+  // iff a matching header exists
+  #include "vast/matching_header.hpp"
+
+  #include "vast/logger.hpp"
+
+  #include <3rd/party.hpp>
+
+  #include <memory>
+
+  #include <sys/types.h>
+  ```
+
+  `clang-format` is configured to automatically change the include order
+  accordingly. Includes separated by preprocessor directives need to be sorted
+  manually.
+
+  Within each section, the order should be alphabetical. VAST includes should
+  always be in double quotes and relative to the source directory, whereas
+  system-wide includes in angle brackets. See below for an example on how to
+  structure includes in unit tests.
+
+- As in the standard library, the order of parameters when declaring a function
+  is: inputs, then outputs. API coherence and symmetry trumps this rule, e.g.,
+  when the first argument of related functions model the same concept.
+
+### Classes
+
+- Use the order `public`, `protected`, `private` for functions and members in
+  classes.
+
+- Mark single-argument constructors as `explicit` to avoid implicit
+  conversions.
```

Comment:
We should add to use `explicit(false)` if a constructor being implicit is actually intended so it's clear that it's not a bug.